src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs

```diff
@@ -0,0 +1,251 @@
+                if (CURLcode.CURLE_OK != result)
+                {
+                    throw new HttpRequestException(SR.net_http_client_execution_error, GetCurlException(result));
+                }
+
+                // TODO: Handle requests with body
+
+                state.CancellationToken.ThrowIfCancellationRequested();
+                responseMessage = CreateResponseMessage(requestHandle, state.RequestMessage);
+            }
+            catch (Exception ex)
+            {
+                savedException = ex;
+            }
+            SafeCurlHandle.DisposeAndClearHandle(ref requestHandle);
+
+            // Move the task to a terminal state.
+            if (responseMessage != null)
+            {
+                state.TrySetResult(responseMessage);
+            }
+            else
+            {
+                HandleAsyncException(state, savedException);
+            }
+        }
+
+        private void SetOperationStarted()
+        {
+            if (!_anyOperationStarted)
+            {
+                _anyOperationStarted = true;
+            }
+        }
+
+        private SafeCurlHandle CreateRequestHandle(RequestState state)
+        {
+            // TODO: If this impacts perf, optimize using a handle pool
+            var requestHandle = Interop.libcurl.curl_easy_init();
+            if (requestHandle.IsInvalid)
+            {
+                throw new HttpRequestException(SR.net_http_client_execution_error);
+            }
+
+            Interop.libcurl.curl_easy_setopt(requestHandle, CURLoption.CURLOPT_URL, state.RequestMessage.RequestUri.OriginalString);
+            if (_automaticRedirection)
+            {
+                Interop.libcurl.curl_easy_setopt(requestHandle, CURLoption.CURLOPT_FOLLOWLOCATION, 1);
+            }
+
+            // TODO: Handle headers, proxy and other options
+
+            return requestHandle;
+        }
+
+
+        private HttpResponseMessage CreateResponseMessage(SafeCurlHandle requestHandle, HttpRequestMessage request)
+        {
+            var response = new HttpResponseMessage();
+            Int64 httpStatusCode = 0;
+            int result = Interop.libcurl.curl_easy_getinfo(requestHandle, CURLINFO.CURLINFO_RESPONSE_CODE, ref httpStatusCode);
+            if (CURLcode.CURLE_OK != result)
+            {
+                throw new HttpRequestException(SR.net_http_client_execution_error, GetCurlException(result));
+            }
+            response.StatusCode = (HttpStatusCode)httpStatusCode;
+
+            // TODO: Do error processing if needed and return actual response
+            response.Content = new StringContent("SendAsync to " + request.RequestUri.OriginalString + " returned: " + response.StatusCode);
+            return response;
+        }
+
+        private void HandleAsyncException(RequestState state, Exception ex)
+        {
+            if (state.CancellationToken.IsCancellationRequested)
```

Comment:
I suggest this if/else be reordered such that exceptions be considered first.  Otherwise, if an exception occurs, and then cancellation is requested, you could end up canceling the operation and dropping the exception on the floor. e.g.
```C#
if (ex != null)
{
    state.TrySetException(ex is HttpRequestException ? ex :
        new HttpRequestException(SR.net_http_client_execution_error, ex));
}
else
{
    Debug.Assert(state.CancellationToken.IsCancellationRequested);
    state.TrySetCanceled(state.CancellationToken);
}
```